### PR TITLE
Update queue with using deque & Update requirements

### PR DESCRIPTION
**Pull request by GLGDLY**
# Description

The performance of queue instance might be improved with the use of double ended queue and `.popleft()`, rather than the list used currently.

Also, for the requirements part, websockets with version<=9.1 would raise `TypeError: WebSocketCommonProtocol.__init__() got an unexpected keyword argument 'logger'`, so it would be better to ensure its version to be >=10.0.

# Checklist:

- [X] I have performed a self-review of my own code
- [X] I have added a short summary of my change to the CHANGELOG.md
- [X] My code follows the style guidelines of this project
- [X] I have commented my code in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by freddyaboulton** [comment]
Thanks @GLGDLY ! We'll test this this week to compare performance with the current queue implementation.

**Comment by GLGDLY** [comment]
> Thanks @GLGDLY ! We'll test this this week to compare performance with the current queue implementation.

For my own massive test considering append, pop, remove and iterate first few items, there is likely to be a 30-40% performance improve comparing deque and list operations. It would be great if you can do some more tests based on actual usage with it, thank you!

**Comment by GLGDLY** [comment]
seems like there are some conflicts now, have you guys test about this deque part? If it is ok then I might resolve the conflicts later @freddyaboulton

**Comment by freddyaboulton** [comment]
Hi @GLGDLY ! Haven't had a chance to test yet. We also want to get #2218 merged which may cause some additional conflicts. I'd suggest we wait until that's merged!

**Comment by GLGDLY** [comment]
> Hi @GLGDLY ! Haven't had a chance to test yet. We also want to get #2218 merged which may cause some additional conflicts. I'd suggest we wait until that's merged!

oh I see, as I have some more stuffs want to discuss and maybe make some more prs, which is not quite related to this pr, so is wondering what's the condition here lol

**Comment by GLGDLY** [comment]
> What do you mean by condition @GLGDLY ?

state? status? I don't quite know if I have used a correct wording🤔

**Comment by freddyaboulton** [comment]
Ah got it! We should be able to review this later this week. Feel free to submit other issues/PRs that are unrelated though! It's just that I know the queue will go through some changes this week (#2218 #2452) so I didn't want you to waste your time addressing changes yet 😄

**Comment by GLGDLY** [comment]
> Ah got it! We should be able to review this later this week. Feel free to submit other issues/PRs that are unrelated though! It's just that I know the queue will go through some changes this week (#2218 #2452) so I didn't want you to waste your time addressing changes yet 😄

ok thanks, I may come and see later😊

**Comment by freddyaboulton** [comment]
Hi @GLGDLY ! If you rebase this pr, we can review it! Thank you

**Comment by GLGDLY** [comment]
> Hi @GLGDLY ! If you rebase this pr, we can review it! Thank you

@freddyaboulton it is now rebased, thank you

File: gradio/queue.py [hunk 1]
```diff
@@ -136,10 +138,11 @@ def push(self, event: Event) -> int | None:
         Returns:
             rank of submitted Event
         """
-        if self.max_size is not None and len(self.event_queue) >= self.max_size:
+        queue_len = len(self.event_queue)
+        if self.max_size is not None and queue_len >= self.max_size:
             return None
         self.event_queue.append(event)
-        return len(self.event_queue) - 1
+        return queue_len
```
Comment (freddyaboulton) [inline gradio/queue.py:145]: Can we keep this as `len(self.event_queue) - 1` please?

**Review by freddyaboulton — COMMENTED** [review]
Thanks for the PR @GLGDLY ! 

I tested this branch against version 3.9 with our queue batch benchmark and there aren't any noticeable regressions in performance.

### Gradio 3.9

| Input type | Average time to complete prediction over 1000 requests |
| ------------- | ------------- |
| Audio  |1.1506862538333997|
| Image  |0.9021098742118249|
| Text |0.8479320884290324 |
| Video |2.458217331471334|


### This branch

| Input type | Average time to complete prediction over 1000 requests |
| ------------- | ------------- |
| Audio  |1.1911795553514513
| Image  |0.9411118013017318|
| Text | 0.9387184648371455|
| Video |2.373436844774655|


I tested locally as well and the time are very comparable. 

The demos are deployed here and they look good to me: https://huggingface.co/spaces/gradio-pr-deploys/pr-2606-all-demos

Any comments before we merge this in @abidlabs ?

**Comment by abidlabs** [comment]
Thanks for benchmarking @freddyaboulton! I was hoping that there would be a more meaningful improvement, but I'm fine with accepting this PR since a deque does seem like the appropriate structure to use here.

**Comment by freddyaboulton** [comment]
Thank you @GLGDLY !

**Comment by GLGDLY** [comment]
> Thank you @GLGDLY ! 

Ok thank you😄

**Pull request merged**